File: docs/create_harel_pdf.py
"""
Script to create Harel Life Insurance Logic PDF documentation
"""

from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib import colors
from reportlab.lib.units import inch
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont

# Create PDF
doc = SimpleDocTemplate(
    "docs/Harel_Life_Insurance_Upload_Logic.pdf",
    pagesize=letter,
    rightMargin=72,
    leftMargin=72,
    topMargin=72,
    bottomMargin=72
)

styles = getSampleStyleSheet()
story = []

# Title
title_style = ParagraphStyle(
    'CustomTitle',
    parent=styles['Title'],
    fontSize=18,
    spaceAfter=30
)
story.append(Paragraph("Harel Life Insurance Upload Logic", title_style))
story.append(Spacer(1, 20))

# Section 1: Overview
story.append(Paragraph("1. Overview", styles['Heading1']))
story.append(Spacer(1, 10))
story.append(Paragraph(
    "This document describes the upload logic for Harel Life Insurance data in the system. "
    "The process involves mapping Excel columns to database fields, parsing agent data, and "
    "storing results in the raw_data table.",
    styles['Normal']
))
story.append(Spacer(1, 20))

# Section 2: Column Mapping
story.append(Paragraph("2. Column Mapping (harelMapping.js)", styles['Heading1']))
story.append(Spacer(1, 10))
story.append(Paragraph("The following Excel columns are mapped for Harel:", styles['Normal']))
story.append(Spacer(1, 10))

# Mapping table
mapping_data = [
    ['Field Name', 'Hebrew Column', 'Description'],
    ['agentName/agentNumber', '__EMPTY (Column A)', 'Agent name and number combined'],
    ['privateRisk', 'sikuni prat', 'Private Risk amount'],
    ['pensionHarel', 'pensioni', 'Pension amount'],
    ['savingsProductsNoFinancials', 'motzrei tzevira lelo finansim', 'Savings without financials'],
    ['pensionTransferNet', 'niyud pensia - neto', 'Pension Transfer Net amount'],
    ['nursingCareHarel', 'nesiot hol', 'Travel Abroad / Nursing Care'],
]

table = Table(mapping_data, colWidths=[2*inch, 2*inch, 2.5*inch])
table.setStyle(TableStyle([
    ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
    ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
    ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
    ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
    ('FONTSIZE', (0, 0), (-1, 0), 10),
    ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
    ('BACKGROUND', (0, 1), (-1, -1), colors.white),
    ('GRID', (0, 0), (-1, -1), 1, colors.black),
    ('FONTSIZE', (0, 1), (-1, -1), 9),
    ('VALIGN', (0, 0), (-1, -1), 'TOP'),
]))
story.append(table)
story.append(Spacer(1, 20))

# Section 3: Data Parsing Logic
story.append(Paragraph("3. Data Parsing Logic (excelParser.js)", styles['Heading1']))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>3.1 Agent Name/Number Extraction</b>", styles['Normal']))
story.append(Spacer(1, 5))
story.append(Paragraph(
    "Harel Excel files contain agent name and number in a combined format in Column A. "
    "The format is: 'Agent Name - Agent Number' (e.g., 'Gal Almagor Agency - 85646'). "
    "The parser splits this using a regex pattern to extract both values.",
    styles['Normal']
))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>3.2 Header Row Detection</b>", styles['Normal']))
story.append(Spacer(1, 5))
story.append(Paragraph(
    "The parser skips header/sub-header rows by checking if numeric columns contain "
    "text like 'tfuka' (output) or 'neto' (net). Only actual data rows are processed.",
    styles['Normal']
))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>3.3 Summary Row Skip</b>", styles['Normal']))
story.append(Spacer(1, 5))
story.append(Paragraph(
    "Rows containing 'sah kol' (total) in the agent name field are skipped as these "
    "are summary rows, not individual agent data.",
    styles['Normal']
))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>3.4 Numeric Validation</b>", styles['Normal']))
story.append(Spacer(1, 5))
story.append(Paragraph(
    "The parser validates that numeric fields (privateRisk, pensionHarel, etc.) contain "
    "actual numbers. Rows where all numeric fields are non-parseable strings are skipped.",
    styles['Normal']
))
story.append(Spacer(1, 20))

# Section 4: Database Storage
story.append(Paragraph("4. Database Storage (uploadRoutes.js)", styles['Heading1']))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>4.1 Target Table</b>", styles['Normal']))
story.append(Spacer(1, 5))
story.append(Paragraph(
    "Life insurance uploads for Harel are stored in the 'raw_data' table.",
    styles['Normal']
))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>4.2 Batch Insertion</b>", styles['Normal']))
story.append(Spacer(1, 5))
story.append(Paragraph(
    "Large files are inserted in batches of 1000 rows to prevent timeout issues.",
    styles['Normal']
))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>4.3 Aggregation</b>", styles['Normal']))
story.append(Spacer(1, 5))
story.append(Paragraph(
    "After successful upload, the aggregation service is triggered to process "
    "and summarize the uploaded data.",
    styles['Normal']
))
story.append(Spacer(1, 20))

# Section 5: Harel-Specific Database Fields
story.append(Paragraph("5. Harel-Specific Database Fields", styles['Heading1']))
story.append(Spacer(1, 10))

fields_data = [
    ['Database Field', 'Source Column', 'Type'],
    ['private_risk', 'sikuni prat', 'Numeric'],
    ['pension_harel', 'pensioni', 'Numeric'],
    ['savings_products_no_financials', 'motzrei tzevira...', 'Numeric'],
    ['pension_transfer_net', 'niyud pensia - neto', 'Numeric'],
    ['nursing_care_harel', 'nesiot hol', 'Numeric'],
]

table2 = Table(fields_data, colWidths=[2.5*inch, 2*inch, 1.5*inch])
table2.setStyle(TableStyle([
    ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
    ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
    ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
    ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
    ('FONTSIZE', (0, 0), (-1, 0), 10),
    ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
    ('BACKGROUND', (0, 1), (-1, -1), colors.white),
    ('GRID', (0, 0), (-1, -1), 1, colors.black),
    ('FONTSIZE', (0, 1), (-1, -1), 9),
]))
story.append(table2)
story.append(Spacer(1, 20))

# Section 6: Files Reference
story.append(Paragraph("6. Source Files Reference", styles['Heading1']))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>harelMapping.js</b> - Column mapping configuration", styles['Normal']))
story.append(Paragraph("Location: gal-almagor-backend/src/config/harelMapping.js", styles['Normal']))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>excelParser.js</b> - Data parsing and transformation logic", styles['Normal']))
story.append(Paragraph("Location: gal-almagor-backend/src/utils/excelParser.js", styles['Normal']))
story.append(Spacer(1, 10))

story.append(Paragraph("<b>uploadRoutes.js</b> - Upload endpoint and database insertion", styles['Normal']))
story.append(Paragraph("Location: gal-almagor-backend/src/routes/uploadRoutes.js", styles['Normal']))

# Build PDF
doc.build(story)
print("PDF created successfully: docs/Harel_Life_Insurance_Upload_Logic.pdf")

File: docs/create_harel_pdf.js
/**
 * Script to create Harel Life Insurance Logic PDF documentation
 */

const PDFDocument = require('pdfkit');
const fs = require('fs');
const path = require('path');

// Create PDF document
const doc = new PDFDocument({
  size: 'LETTER',
  margins: { top: 72, bottom: 72, left: 72, right: 72 }
});

// Output file
const outputPath = path.join(__dirname, 'Harel_Life_Insurance_Upload_Logic.pdf');
doc.pipe(fs.createWriteStream(outputPath));

// Title
doc.fontSize(20).font('Helvetica-Bold')
   .text('Harel Life Insurance Upload Logic', { align: 'center' });
doc.moveDown(2);

// Section 1: Overview
doc.fontSize(14).font('Helvetica-Bold').text('1. Overview');
doc.moveDown(0.5);
doc.fontSize(11).font('Helvetica')
   .text('This document describes the upload logic for Harel Life Insurance data in the system. The process involves mapping Excel columns to database fields, parsing agent data, and storing results in the raw_data table.', {
     align: 'justify'
   });
doc.moveDown(1.5);

// Section 2: Column Mapping
doc.fontSize(14).font('Helvetica-Bold').text('2. Column Mapping (harelMapping.js)');
doc.moveDown(0.5);
doc.fontSize(11).font('Helvetica')
   .text('The following Excel columns are mapped for Harel:');
doc.moveDown(0.5);

// Mapping table data
const mappings = [
  { field: 'agentName/agentNumber', hebrew: '__EMPTY (Column A)', desc: 'Agent name and number combined' },
  { field: 'privateRisk', hebrew: 'סיכוני פרט', desc: 'Private Risk amount' },
  { field: 'pensionHarel', hebrew: 'פנסיוני', desc: 'Pension amount' },
  { field: 'savingsProductsNoFinancials', hebrew: 'מוצרי צבירה ללא פיננסים', desc: 'Savings without financials' },
  { field: 'pensionTransferNet', hebrew: 'ניוד פנסיה - נטו', desc: 'Pension Transfer Net' },
  { field: 'nursingCareHarel', hebrew: 'נסיעות חול', desc: 'Travel Abroad / Nursing Care' }
];

// Draw table
const tableTop = doc.y;
const col1X = 72;
const col2X = 220;
const col3X = 370;
const rowHeight = 25;

// Header row
doc.rect(col1X, tableTop, 468, rowHeight).fill('#666666');
doc.fillColor('white').fontSize(10).font('Helvetica-Bold');
doc.text('Field Name', col1X + 5, tableTop + 7, { width: 140 });
doc.text('Hebrew Column', col2X + 5, tableTop + 7, { width: 140 });
doc.text('Description', col3X + 5, tableTop + 7, { width: 160 });

// Data rows
doc.fillColor('black').font('Helvetica').fontSize(9);
mappings.forEach((row, i) => {
  const y = tableTop + rowHeight + (i * rowHeight);
  doc.rect(col1X, y, 468, rowHeight).stroke();
  doc.text(row.field, col1X + 5, y + 7, { width: 140 });
  doc.text(row.hebrew, col2X + 5, y + 7, { width: 140 });
  doc.text(row.desc, col3X + 5, y + 7, { width: 160 });
});

doc.y = tableTop + rowHeight + (mappings.length * rowHeight) + 20;
doc.moveDown(1);

// Section 3: Data Parsing Logic
doc.fontSize(14).font('Helvetica-Bold').text('3. Data Parsing Logic (excelParser.js)');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('3.1 Agent Name/Number Extraction');
doc.fontSize(10).font('Helvetica')
   .text('Harel Excel files contain agent name and number in a combined format in Column A. The format is: "Agent Name - Agent Number" (e.g., "גל אלמגור-דאוד סוכנות - 85646"). The parser splits this using a regex pattern to extract both values.');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('3.2 Header Row Detection');
doc.fontSize(10).font('Helvetica')
   .text('The parser skips header/sub-header rows by checking if numeric columns contain text like "תפוקה" (output) or "נטו" (net). Only actual data rows are processed.');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('3.3 Summary Row Skip');
doc.fontSize(10).font('Helvetica')
   .text('Rows containing "סה״כ" (total) in the agent name field are skipped as these are summary rows, not individual agent data.');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('3.4 Numeric Validation');
doc.fontSize(10).font('Helvetica')
   .text('The parser validates that numeric fields (privateRisk, pensionHarel, etc.) contain actual numbers. Rows where all numeric fields are non-parseable strings are skipped.');
doc.moveDown(1.5);

// Section 4: Database Storage
doc.fontSize(14).font('Helvetica-Bold').text('4. Database Storage (uploadRoutes.js)');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('4.1 Target Table');
doc.fontSize(10).font('Helvetica')
   .text('Life insurance uploads for Harel are stored in the "raw_data" table.');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('4.2 Batch Insertion');
doc.fontSize(10).font('Helvetica')
   .text('Large files are inserted in batches of 1000 rows to prevent timeout issues.');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('4.3 Aggregation');
doc.fontSize(10).font('Helvetica')
   .text('After successful upload, the aggregation service is triggered to process and summarize the uploaded data.');
doc.moveDown(1.5);

// Section 5: Database Fields
doc.fontSize(14).font('Helvetica-Bold').text('5. Harel-Specific Database Fields');
doc.moveDown(0.5);

const dbFields = [
  { field: 'private_risk', source: 'סיכוני פרט', type: 'Numeric' },
  { field: 'pension_harel', source: 'פנסיוני', type: 'Numeric' },
  { field: 'savings_products_no_financials', source: 'מוצרי צבירה ללא פיננסים', type: 'Numeric' },
  { field: 'pension_transfer_net', source: 'ניוד פנסיה - נטו', type: 'Numeric' },
  { field: 'nursing_care_harel', source: 'נסיעות חול', type: 'Numeric' }
];

// Draw second table
const table2Top = doc.y;

// Header row
doc.rect(col1X, table2Top, 468, rowHeight).fill('#666666');
doc.fillColor('white').fontSize(10).font('Helvetica-Bold');
doc.text('Database Field', col1X + 5, table2Top + 7, { width: 180 });
doc.text('Source Column', col2X + 5, table2Top + 7, { width: 150 });
doc.text('Type', col3X + 60, table2Top + 7, { width: 80 });

// Data rows
doc.fillColor('black').font('Helvetica').fontSize(9);
dbFields.forEach((row, i) => {
  const y = table2Top + rowHeight + (i * rowHeight);
  doc.rect(col1X, y, 468, rowHeight).stroke();
  doc.text(row.field, col1X + 5, y + 7, { width: 180 });
  doc.text(row.source, col2X + 5, y + 7, { width: 150 });
  doc.text(row.type, col3X + 60, y + 7, { width: 80 });
});

doc.y = table2Top + rowHeight + (dbFields.length * rowHeight) + 20;
doc.moveDown(1);

// Section 6: Files Reference
doc.fontSize(14).font('Helvetica-Bold').text('6. Source Files Reference');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('harelMapping.js');
doc.fontSize(10).font('Helvetica').text('Column mapping configuration');
doc.fontSize(9).fillColor('gray').text('Location: gal-almagor-backend/src/config/harelMapping.js');
doc.fillColor('black');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('excelParser.js');
doc.fontSize(10).font('Helvetica').text('Data parsing and transformation logic');
doc.fontSize(9).fillColor('gray').text('Location: gal-almagor-backend/src/utils/excelParser.js');
doc.fillColor('black');
doc.moveDown(0.5);

doc.fontSize(11).font('Helvetica-Bold').text('uploadRoutes.js');
doc.fontSize(10).font('Helvetica').text('Upload endpoint and database insertion');
doc.fontSize(9).fillColor('gray').text('Location: gal-almagor-backend/src/routes/uploadRoutes.js');

// Finalize PDF
doc.end();

console.log('PDF created successfully:', outputPath);
